man/zmirror.1
.TH ZMIRROR "1" "2025\-09\-10" "zmirror" "Generated Python Manual"
.SH NAME
zmirror
.SH SYNOPSIS
.B zmirror
[-h] [--version] {daemon,clear-cache,reload-config,scrub-all,scrub-overdue,resilver-overdue,trim-all,trim-overdue,online-all,status-all,daemon-version,maintenance,online,offline,status,trim,scrub,list,enable,disable,set,get} ...

.SH
POSITIONAL ARGUMENTS
.TP
\fBzmirror\fR \fI\,daemon\/\fR
starts zmirror in daemon mode
.TP
\fBzmirror\fR \fI\,clear\-cache\/\fR
clears the cache and removes the cache file (which stores the dates of when the maintenance tasks were last run). Triggers a configuration reload.
.TP
\fBzmirror\fR \fI\,reload\-config\/\fR
reloads the configuration.
.TP
\fBzmirror\fR \fI\,scrub\-all\/\fR
requests all configured zdevs to be scrubbed. This will bring all necessary and available parent devices (i.e. dm\-crypts) online.
.TP
\fBzmirror\fR \fI\,scrub\-overdue\/\fR
requests zdevs to be scrubbed if they are behind their configured `scrub_interval`. This will bring all necessary and available parent devices (i.e. `dm\-crypt`s) online.
.TP
\fBzmirror\fR \fI\,resilver\-overdue\/\fR
requests zdevs to be resilvered if they are behind their configured `resilver_interval`. Since a resilver happens whenever a mirrored device is brought online, this really does nothing but (try to) online the respective devices.
.TP
\fBzmirror\fR \fI\,trim\-all\/\fR
requests all configured zdevs to be trimmed. This will bring all necessary and available parent devices (i.e. `dm\-crypt`s) online.
.TP
\fBzmirror\fR \fI\,trim\-overdue\/\fR
requests zdevs to be trimmed if they are behind their configured trim_interval. This will bring all necessary and available parent devices (i.e. `dm\-crypt`s) online.
.TP
\fBzmirror\fR \fI\,online\-all\/\fR
requests all configured devices to be onlined.
.TP
\fBzmirror\fR \fI\,status\-all\/\fR
shows the status of all configured devices
.TP
\fBzmirror\fR \fI\,daemon\-version\/\fR
shows the version of the (currently running) zmirror daemon
.TP
\fBzmirror\fR \fI\,maintenance\/\fR
triggers all maintenance tasks (scrub and trim as scheduled as well as onlining devices that should be brought up to date [resilvered]). This is what you want to schedule via a cronjob or a systemd timer. It makes sense to do this at night on a weekday where there is not much load on your machine. This will online all devices that are present and need maintenance. The devices will only be taken offline afterwards if you have configured the respective event timers to take them offline (i.e. `on_resilvered` set to `offline`).
.TP
\fBzmirror\fR \fI\,online\/\fR
request device to go online
.TP
\fBzmirror\fR \fI\,offline\/\fR
request device to go offline
.TP
\fBzmirror\fR \fI\,status\/\fR
show device status
.TP
\fBzmirror\fR \fI\,trim\/\fR
request device trim
.TP
\fBzmirror\fR \fI\,scrub\/\fR
request device to be scrubbed
.TP
\fBzmirror\fR \fI\,list\/\fR
list devices fulfilling conditions
.TP
\fBzmirror\fR \fI\,enable\/\fR
enable zmirror daemon property
.TP
\fBzmirror\fR \fI\,disable\/\fR
disable zmirror daemon property
.TP
\fBzmirror\fR \fI\,set\/\fR
set zmirror daemon property to value
.TP
\fBzmirror\fR \fI\,get\/\fR
get zmirror daemon property value

.SH COMMAND \fI\,'zmirror daemon'\/\fR
usage: zmirror daemon [\-h] [\-\-config\-path CONFIG_PATH]
                      [\-\-cache\-path CACHE_PATH] [\-\-socket\-path SOCKET_PATH]

.SH OPTIONS \fI\,'zmirror daemon'\/\fR
.TP
\fB\-\-config\-path\fR \fI\,CONFIG_PATH\/\fR
the path to the config file. May be set via the environment variable
`ZMIRROR_CONFIG_PATH`. Defaults to `/etc/zmirror/zmirror\-config.yml`

.TP
\fB\-\-cache\-path\fR \fI\,CACHE_PATH\/\fR
the path to the cache file in which zmirror stores device state. May be set
via the environment variable `ZMIRROR_CACHE_PATH`. Defaults to
`/var/lib/zmirror/zmirror\-cache.yml`

.TP
\fB\-\-socket\-path\fR \fI\,SOCKET_PATH\/\fR
the path to the unix socket on which `zmirror daemon` listens, to which
`zmirror\-trigger` sends UDEV and ZED events, and to which commands are sent
when you invoke `zmirror <command>`. May be set via the environment variable
`ZMIRROR_SOCKET_PATH`. Defaults to `/run/zmirror/zmirror.socket`

.SH COMMAND \fI\,'zmirror clear\-cache'\/\fR
usage: zmirror clear\-cache [\-h] [\-\-socket\-path SOCKET_PATH]

.SH OPTIONS \fI\,'zmirror clear\-cache'\/\fR
.TP
\fB\-\-socket\-path\fR \fI\,SOCKET_PATH\/\fR
the path to the unix socket on which `zmirror daemon` listens, to which
`zmirror\-trigger` sends UDEV and ZED events, and to which commands are sent
when you invoke `zmirror <command>`. May be set via the environment variable
`ZMIRROR_SOCKET_PATH`. Defaults to `/run/zmirror/zmirror.socket`

.SH COMMAND \fI\,'zmirror reload\-config'\/\fR
usage: zmirror reload\-config [\-h] [\-\-socket\-path SOCKET_PATH]

.SH OPTIONS \fI\,'zmirror reload\-config'\/\fR
.TP
\fB\-\-socket\-path\fR \fI\,SOCKET_PATH\/\fR
the path to the unix socket on which `zmirror daemon` listens, to which
`zmirror\-trigger` sends UDEV and ZED events, and to which commands are sent
when you invoke `zmirror <command>`. May be set via the environment variable
`ZMIRROR_SOCKET_PATH`. Defaults to `/run/zmirror/zmirror.socket`

.SH COMMAND \fI\,'zmirror scrub\-all'\/\fR
usage: zmirror scrub\-all [\-h] [\-\-socket\-path SOCKET_PATH] [\-\-cancel]

.SH OPTIONS \fI\,'zmirror scrub\-all'\/\fR
.TP
\fB\-\-socket\-path\fR \fI\,SOCKET_PATH\/\fR
the path to the unix socket on which `zmirror daemon` listens, to which
`zmirror\-trigger` sends UDEV and ZED events, and to which commands are sent
when you invoke `zmirror <command>`. May be set via the environment variable
`ZMIRROR_SOCKET_PATH`. Defaults to `/run/zmirror/zmirror.socket`

.TP
\fB\-\-cancel\fR

.SH COMMAND \fI\,'zmirror scrub\-overdue'\/\fR
usage: zmirror scrub\-overdue [\-h] [\-\-socket\-path SOCKET_PATH]

.SH OPTIONS \fI\,'zmirror scrub\-overdue'\/\fR
.TP
\fB\-\-socket\-path\fR \fI\,SOCKET_PATH\/\fR
the path to the unix socket on which `zmirror daemon` listens, to which
`zmirror\-trigger` sends UDEV and ZED events, and to which commands are sent
when you invoke `zmirror <command>`. May be set via the environment variable
`ZMIRROR_SOCKET_PATH`. Defaults to `/run/zmirror/zmirror.socket`

.SH COMMAND \fI\,'zmirror resilver\-overdue'\/\fR
usage: zmirror resilver\-overdue [\-h] [\-\-socket\-path SOCKET_PATH]

.SH OPTIONS \fI\,'zmirror resilver\-overdue'\/\fR
.TP
\fB\-\-socket\-path\fR \fI\,SOCKET_PATH\/\fR
the path to the unix socket on which `zmirror daemon` listens, to which
`zmirror\-trigger` sends UDEV and ZED events, and to which commands are sent
when you invoke `zmirror <command>`. May be set via the environment variable
`ZMIRROR_SOCKET_PATH`. Defaults to `/run/zmirror/zmirror.socket`

.SH COMMAND \fI\,'zmirror trim\-all'\/\fR
usage: zmirror trim\-all [\-h] [\-\-socket\-path SOCKET_PATH] [\-\-cancel]

.SH OPTIONS \fI\,'zmirror trim\-all'\/\fR
.TP
\fB\-\-socket\-path\fR \fI\,SOCKET_PATH\/\fR
the path to the unix socket on which `zmirror daemon` listens, to which
`zmirror\-trigger` sends UDEV and ZED events, and to which commands are sent
when you invoke `zmirror <command>`. May be set via the environment variable
`ZMIRROR_SOCKET_PATH`. Defaults to `/run/zmirror/zmirror.socket`

.TP
\fB\-\-cancel\fR

.SH COMMAND \fI\,'zmirror trim\-overdue'\/\fR
usage: zmirror trim\-overdue [\-h] [\-\-socket\-path SOCKET_PATH]

.SH OPTIONS \fI\,'zmirror trim\-overdue'\/\fR
.TP
\fB\-\-socket\-path\fR \fI\,SOCKET_PATH\/\fR
the path to the unix socket on which `zmirror daemon` listens, to which
`zmirror\-trigger` sends UDEV and ZED events, and to which commands are sent
when you invoke `zmirror <command>`. May be set via the environment variable
`ZMIRROR_SOCKET_PATH`. Defaults to `/run/zmirror/zmirror.socket`

.SH COMMAND \fI\,'zmirror online\-all'\/\fR
usage: zmirror online\-all [\-h] [\-\-socket\-path SOCKET_PATH] [\-\-cancel]

.SH OPTIONS \fI\,'zmirror online\-all'\/\fR
.TP
\fB\-\-socket\-path\fR \fI\,SOCKET_PATH\/\fR
the path to the unix socket on which `zmirror daemon` listens, to which
`zmirror\-trigger` sends UDEV and ZED events, and to which commands are sent
when you invoke `zmirror <command>`. May be set via the environment variable
`ZMIRROR_SOCKET_PATH`. Defaults to `/run/zmirror/zmirror.socket`

.TP
\fB\-\-cancel\fR

.SH COMMAND \fI\,'zmirror status\-all'\/\fR
usage: zmirror status\-all [\-h] [\-\-socket\-path SOCKET_PATH]

.SH OPTIONS \fI\,'zmirror status\-all'\/\fR
.TP
\fB\-\-socket\-path\fR \fI\,SOCKET_PATH\/\fR
the path to the unix socket on which `zmirror daemon` listens, to which
`zmirror\-trigger` sends UDEV and ZED events, and to which commands are sent
when you invoke `zmirror <command>`. May be set via the environment variable
`ZMIRROR_SOCKET_PATH`. Defaults to `/run/zmirror/zmirror.socket`

.SH COMMAND \fI\,'zmirror daemon\-version'\/\fR
usage: zmirror daemon\-version [\-h] [\-\-socket\-path SOCKET_PATH]

.SH OPTIONS \fI\,'zmirror daemon\-version'\/\fR
.TP
\fB\-\-socket\-path\fR \fI\,SOCKET_PATH\/\fR
the path to the unix socket on which `zmirror daemon` listens, to which
`zmirror\-trigger` sends UDEV and ZED events, and to which commands are sent
when you invoke `zmirror <command>`. May be set via the environment variable
`ZMIRROR_SOCKET_PATH`. Defaults to `/run/zmirror/zmirror.socket`

.SH COMMAND \fI\,'zmirror maintenance'\/\fR
usage: zmirror maintenance [\-h] [\-\-socket\-path SOCKET_PATH]

.SH OPTIONS \fI\,'zmirror maintenance'\/\fR
.TP
\fB\-\-socket\-path\fR \fI\,SOCKET_PATH\/\fR
the path to the unix socket on which `zmirror daemon` listens, to which
`zmirror\-trigger` sends UDEV and ZED events, and to which commands are sent
when you invoke `zmirror <command>`. May be set via the environment variable
`ZMIRROR_SOCKET_PATH`. Defaults to `/run/zmirror/zmirror.socket`

.SH COMMAND \fI\,'zmirror online'\/\fR
usage: zmirror online [\-h] [\-\-socket\-path SOCKET_PATH] [\-\-cancel]
                      {disk,partition,zpool,zfs\-volume,dm\-crypt,zdev} ...

.SH
POSITIONAL ARGUMENTS \fI\,'zmirror online'\/\fR


.SH COMMAND \fI\,'zmirror online disk'\/\fR
usage: zmirror online disk [\-h] [\-\-cancel] uuid

.TP
\fBuuid\fR
id field

.SH OPTIONS \fI\,'zmirror online disk'\/\fR
.TP
\fB\-\-cancel\fR

.SH COMMAND \fI\,'zmirror online partition'\/\fR
usage: zmirror online partition [\-h] [\-\-cancel] name

.TP
\fBname\fR
id field

.SH OPTIONS \fI\,'zmirror online partition'\/\fR
.TP
\fB\-\-cancel\fR

.SH COMMAND \fI\,'zmirror online zpool'\/\fR
usage: zmirror online zpool [\-h] [\-\-cancel] name

.TP
\fBname\fR
id field

.SH OPTIONS \fI\,'zmirror online zpool'\/\fR
.TP
\fB\-\-cancel\fR

.SH COMMAND \fI\,'zmirror online zfs\-volume'\/\fR
usage: zmirror online zfs\-volume [\-h] [\-\-cancel] pool name

.TP
\fBpool\fR
id field

.TP
\fBname\fR
id field

.SH OPTIONS \fI\,'zmirror online zfs\-volume'\/\fR
.TP
\fB\-\-cancel\fR

.SH COMMAND \fI\,'zmirror online dm\-crypt'\/\fR
usage: zmirror online dm\-crypt [\-h] [\-\-cancel] name

.TP
\fBname\fR
id field

.SH OPTIONS \fI\,'zmirror online dm\-crypt'\/\fR
.TP
\fB\-\-cancel\fR

.SH COMMAND \fI\,'zmirror online zdev'\/\fR
usage: zmirror online zdev [\-h] [\-\-cancel] pool name

.TP
\fBpool\fR
id field

.TP
\fBname\fR
id field

.SH OPTIONS \fI\,'zmirror online zdev'\/\fR
.TP
\fB\-\-cancel\fR

.SH OPTIONS \fI\,'zmirror online'\/\fR
.TP
\fB\-\-socket\-path\fR \fI\,SOCKET_PATH\/\fR
the path to the unix socket on which `zmirror daemon` listens, to which
`zmirror\-trigger` sends UDEV and ZED events, and to which commands are sent
when you invoke `zmirror <command>`. May be set via the environment variable
`ZMIRROR_SOCKET_PATH`. Defaults to `/run/zmirror/zmirror.socket`

.TP
\fB\-\-cancel\fR

.SH COMMAND \fI\,'zmirror offline'\/\fR
usage: zmirror offline [\-h] [\-\-socket\-path SOCKET_PATH]
                       {disk,partition,zpool,zfs\-volume,dm\-crypt,zdev} ...

.SH
POSITIONAL ARGUMENTS \fI\,'zmirror offline'\/\fR


.SH COMMAND \fI\,'zmirror offline disk'\/\fR
usage: zmirror offline disk [\-h] [\-\-cancel] uuid

.TP
\fBuuid\fR
id field

.SH OPTIONS \fI\,'zmirror offline disk'\/\fR
.TP
\fB\-\-cancel\fR

.SH COMMAND \fI\,'zmirror offline partition'\/\fR
usage: zmirror offline partition [\-h] [\-\-cancel] name

.TP
\fBname\fR
id field

.SH OPTIONS \fI\,'zmirror offline partition'\/\fR
.TP
\fB\-\-cancel\fR

.SH COMMAND \fI\,'zmirror offline zpool'\/\fR
usage: zmirror offline zpool [\-h] [\-\-cancel] name

.TP
\fBname\fR
id field

.SH OPTIONS \fI\,'zmirror offline zpool'\/\fR
.TP
\fB\-\-cancel\fR

.SH COMMAND \fI\,'zmirror offline zfs\-volume'\/\fR
usage: zmirror offline zfs\-volume [\-h] [\-\-cancel] pool name

.TP
\fBpool\fR
id field

.TP
\fBname\fR
id field

.SH OPTIONS \fI\,'zmirror offline zfs\-volume'\/\fR
.TP
\fB\-\-cancel\fR

.SH COMMAND \fI\,'zmirror offline dm\-crypt'\/\fR
usage: zmirror offline dm\-crypt [\-h] [\-\-cancel] name

.TP
\fBname\fR
id field

.SH OPTIONS \fI\,'zmirror offline dm\-crypt'\/\fR
.TP
\fB\-\-cancel\fR

.SH COMMAND \fI\,'zmirror offline zdev'\/\fR
usage: zmirror offline zdev [\-h] [\-\-cancel] pool name

.TP
\fBpool\fR
id field

.TP
\fBname\fR
id field

.SH OPTIONS \fI\,'zmirror offline zdev'\/\fR
.TP
\fB\-\-cancel\fR

.SH OPTIONS \fI\,'zmirror offline'\/\fR
.TP
\fB\-\-socket\-path\fR \fI\,SOCKET_PATH\/\fR
the path to the unix socket on which `zmirror daemon` listens, to which
`zmirror\-trigger` sends UDEV and ZED events, and to which commands are sent
when you invoke `zmirror <command>`. May be set via the environment variable
`ZMIRROR_SOCKET_PATH`. Defaults to `/run/zmirror/zmirror.socket`

.SH COMMAND \fI\,'zmirror status'\/\fR
usage: zmirror status [\-h] [\-\-socket\-path SOCKET_PATH]
                      {disk,partition,zpool,zfs\-volume,dm\-crypt,zdev} ...

.SH
POSITIONAL ARGUMENTS \fI\,'zmirror status'\/\fR


.SH COMMAND \fI\,'zmirror status disk'\/\fR
usage: zmirror status disk [\-h] uuid

.TP
\fBuuid\fR
id field

.SH COMMAND \fI\,'zmirror status partition'\/\fR
usage: zmirror status partition [\-h] name

.TP
\fBname\fR
id field

.SH COMMAND \fI\,'zmirror status zpool'\/\fR
usage: zmirror status zpool [\-h] name

.TP
\fBname\fR
id field

.SH COMMAND \fI\,'zmirror status zfs\-volume'\/\fR
usage: zmirror status zfs\-volume [\-h] pool name

.TP
\fBpool\fR
id field

.TP
\fBname\fR
id field

.SH COMMAND \fI\,'zmirror status dm\-crypt'\/\fR
usage: zmirror status dm\-crypt [\-h] name

.TP
\fBname\fR
id field

.SH COMMAND \fI\,'zmirror status zdev'\/\fR
usage: zmirror status zdev [\-h] pool name

.TP
\fBpool\fR
id field

.TP
\fBname\fR
id field

.SH OPTIONS \fI\,'zmirror status'\/\fR
.TP
\fB\-\-socket\-path\fR \fI\,SOCKET_PATH\/\fR
the path to the unix socket on which `zmirror daemon` listens, to which
`zmirror\-trigger` sends UDEV and ZED events, and to which commands are sent
when you invoke `zmirror <command>`. May be set via the environment variable
`ZMIRROR_SOCKET_PATH`. Defaults to `/run/zmirror/zmirror.socket`

.SH COMMAND \fI\,'zmirror trim'\/\fR
usage: zmirror trim [\-h] [\-\-socket\-path SOCKET_PATH] {zdev} ...

.SH
POSITIONAL ARGUMENTS \fI\,'zmirror trim'\/\fR


.SH COMMAND \fI\,'zmirror trim zdev'\/\fR
usage: zmirror trim zdev [\-h] [\-\-cancel] pool name

.TP
\fBpool\fR
id field

.TP
\fBname\fR
id field

.SH OPTIONS \fI\,'zmirror trim zdev'\/\fR
.TP
\fB\-\-cancel\fR

.SH OPTIONS \fI\,'zmirror trim'\/\fR
.TP
\fB\-\-socket\-path\fR \fI\,SOCKET_PATH\/\fR
the path to the unix socket on which `zmirror daemon` listens, to which
`zmirror\-trigger` sends UDEV and ZED events, and to which commands are sent
when you invoke `zmirror <command>`. May be set via the environment variable
`ZMIRROR_SOCKET_PATH`. Defaults to `/run/zmirror/zmirror.socket`

.SH COMMAND \fI\,'zmirror scrub'\/\fR
usage: zmirror scrub [\-h] [\-\-socket\-path SOCKET_PATH] [\-\-cancel]
                     {zpool,zdev} ...

.SH
POSITIONAL ARGUMENTS \fI\,'zmirror scrub'\/\fR


.SH COMMAND \fI\,'zmirror scrub zpool'\/\fR
usage: zmirror scrub zpool [\-h] name

.TP
\fBname\fR
id field

.SH COMMAND \fI\,'zmirror scrub zdev'\/\fR
usage: zmirror scrub zdev [\-h] pool name

.TP
\fBpool\fR
id field

.TP
\fBname\fR
id field

.SH OPTIONS \fI\,'zmirror scrub'\/\fR
.TP
\fB\-\-socket\-path\fR \fI\,SOCKET_PATH\/\fR
the path to the unix socket on which `zmirror daemon` listens, to which
`zmirror\-trigger` sends UDEV and ZED events, and to which commands are sent
when you invoke `zmirror <command>`. May be set via the environment variable
`ZMIRROR_SOCKET_PATH`. Defaults to `/run/zmirror/zmirror.socket`

.TP
\fB\-\-cancel\fR

.SH COMMAND \fI\,'zmirror list'\/\fR
usage: zmirror list [\-h] [\-\-socket\-path SOCKET_PATH] [\-\-keys KEYS]
                    [\-\-no_headers]
                    [\-\-format {json,plain,simple,github,grid,simple_grid,rounded_grid,heavy_grid,mixed_grid,double_grid,fancy_grid,outline,simple_outline,rounded_outline,heavy_outline,mixed_outline,double_outline,fancy_outline,pipe,orgtbl,asciidoc,jira,presto,pretty,psql,rst,mediawiki,moinmoin,youtrack,html,unsafehtml,latex,latex_raw,latex_booktabs,latex_longtable,textile,tsv}]
                    [\-\-sort {id,last_online,last_update,update_overdue,last_trim,trim_overdue,last_scrub,scrub_overdue}]
                    {overdue,scrub,trim,update} ...

.SH
POSITIONAL ARGUMENTS \fI\,'zmirror list'\/\fR
.TP
\fBzmirror list\fR \fI\,overdue\/\fR
list all overdue devices
.TP
\fBzmirror list\fR \fI\,scrub\/\fR
list devices based on current/last scrub state
.TP
\fBzmirror list\fR \fI\,trim\/\fR
list devices based on current/last trim state
.TP
\fBzmirror list\fR \fI\,update\/\fR
list devices based on when they were last up\-to\-date

.SH COMMAND \fI\,'zmirror list overdue'\/\fR
usage: zmirror list overdue [\-h] [\-\-keys KEYS] [\-\-no_headers]

zmirror/zmirror.py




import traceback
import sys
import argparse


from .logging import log
from .dataclasses import *
from .util import myexec, outs, copy_attrs, env_var_or
from .entities import *
from . import commands as commands
from .daemon import daemon
from kpyutils.kiify import KdStream, date_or_datetime

from .user_commands import *

from . import operations as operations

from .defaults import *


from ._version import __version__


def make_arg_parser():

  parser = argparse.ArgumentParser(prog="zmirror")
  parser.add_argument('--version', action='version', version=__version__, help="print the version")
  subs = parser.add_subparsers(required=True)


  # nice to have, maybe?: zmirror trigger        # simulates events or something...


  # nice to have: zmirror prune-cache    # called by user
  # prune_parser = subs.add_parser('prune-cache', parents=[], help='remove cache entries that are not also present in config')
  # prune_parser.set_defaults(func=prune_cache)

  shared_parser = argparse.ArgumentParser(add_help=False)

  shared_parser.add_argument("--config-path", type=str, help=f"the path to the config file. May be set via the environment variable `ZMIRROR_CONFIG_PATH`. Defaults to `{ZMIRROR_CONFIG_PATH_DEFAULT}`", default=env_var_or("ZMIRROR_CONFIG_PATH", ZMIRROR_CONFIG_PATH_DEFAULT))

  shared_parser.add_argument("--cache-path", type=str, help=f"the path to the cache file in which zmirror stores device state. May be set via the environment variable `ZMIRROR_CACHE_PATH`. Defaults to `{ZMIRROR_CACHE_PATH_DEFAULT}`", default=env_var_or("ZMIRROR_CACHE_PATH", ZMIRROR_CACHE_PATH_DEFAULT))

  socket_parser = argparse.ArgumentParser(add_help=False)
  socket_parser.add_argument("--socket-path", type=str, help=f"the path to the unix socket on which `zmirror daemon` listens, to which `zmirror-trigger` sends UDEV and ZED events, and to which commands are sent when you invoke `zmirror <command>`. May be set via the environment variable `ZMIRROR_SOCKET_PATH`. Defaults to `{ZMIRROR_SOCKET_PATH_DEFAULT}`", default=env_var_or("ZMIRROR_SOCKET_PATH", ZMIRROR_SOCKET_PATH_DEFAULT))

  # shared_parser.add_argument("runtime-dir", type=str, help="the path to the runtime directory", default= "/var/run/zmirror")

  cancel_parser = argparse.ArgumentParser(add_help=False)
  cancel_parser.add_argument("--cancel", action="store_true")


  list_args_parser = argparse.ArgumentParser(add_help=False)


  TABLE_FORMATS = [
    "json",

    # tabulate's formats
    "plain",
    "simple",
    "github",
    "grid",
    "simple_grid",
    "rounded_grid",
    "heavy_grid",
    "mixed_grid",
    "double_grid",
    "fancy_grid",
    "outline",
    "simple_outline",
    "rounded_outline",
    "heavy_outline",
    "mixed_outline",
    "double_outline",
    "fancy_outline",
    "pipe",
    "orgtbl",
    "asciidoc",
    "jira",
    "presto",
    "pretty",
    "psql",
    "rst",
    "mediawiki",
    "moinmoin",
    "youtrack",
    "html",
    "unsafehtml",
    "latex",
    "latex_raw",
    "latex_booktabs",
    "latex_longtable",
    "textile",
    "tsv"
  ]


  list_args_parser.add_argument("--keys", type=lambda s: s.split(','), default=LIST_KEYS, help="only output this list of keys (columns)")
  list_args_parser.add_argument("--no_headers", action="store_true", default=False, help="do not print headers when outputting a table")
  list_args_parser.add_argument("--format", choices=TABLE_FORMATS, default="plain", help="either `json` or one of the formats defined by the tabulate library (see https://https://pypi.org/project/tabulate/#description)")
  list_args_parser.add_argument("--sort", choices=LIST_KEYS, help="the key (column) to sort for")

  daemon_parser = subs.add_parser('daemon', parents=[shared_parser, socket_parser], help="starts zmirror in daemon mode")
  daemon_parser.set_defaults(func=daemon)


  # daemon commands
  # #######################


  def subcmd(command, help=None, cancel=False):
    parents = [socket_parser]
    if cancel:
      parents.append(cancel_parser)
    subpsr = subs.add_parser(command, parents=parents, help=help)
    subpsr.set_defaults(func=make_send_simple_daemon_command(command))


  subcmd("clear-cache", help="clears the cache and removes the cache file (which stores the dates of when the maintenance tasks were last run). Triggers a configuration reload.")
  subcmd("reload-config", help="reloads the configuration.")

  subcmd("scrub-all", cancel=True, help="requests all configured zdevs to be scrubbed. This will bring all necessary and available parent devices (i.e. dm-crypts) online.")
  subcmd("scrub-overdue", help="requests zdevs to be scrubbed if they are behind their configured `scrub_interval`. This will bring all necessary and available parent devices (i.e. `dm-crypt`s) online.")

  subcmd("resilver-overdue", help="requests zdevs to be resilvered if they are behind their configured `resilver_interval`. Since a resilver happens whenever a mirrored device is brought online, this really does nothing but (try to) online the respective devices.")

  subcmd("trim-all", cancel=True, help="requests all configured zdevs to be trimmed. This will bring all necessary and available parent devices (i.e. `dm-crypt`s) online.")
  subcmd("trim-overdue", help="requests zdevs to be trimmed if they are behind their configured trim_interval. This will bring all necessary and available parent devices (i.e. `dm-crypt`s) online.")


  subcmd("online-all", cancel=True, help="requests all configured devices to be onlined.")

  subcmd("status-all", help="shows the status of all configured devices")
  subcmd("daemon-version", help="shows the version of the (currently running) zmirror daemon")

  subcmd("maintenance", help="triggers all maintenance tasks (scrub and trim as scheduled as well as onlining devices that should be brought up to date [resilvered]). This is what you want to schedule via a cronjob or a systemd timer. It makes sense to do this at night on a weekday where there is not much load on your machine. This will online all devices that are present and need maintenance. The devices will only be taken offline afterwards if you have configured the respective event timers to take them offline (i.e. `on_resilvered` set to `offline`).")


  # scrub_parser = subs.add_parser('scrub-overdue', parents=[], help='scrub devices that have not been scrubbed for too long')
  # scrub_parser.set_defaults(func=scrub)

  online_parser = subs.add_parser('online', parents=[socket_parser, cancel_parser], help='request device to go online')
  offline_parser = subs.add_parser('offline', parents=[socket_parser], help='request device to go offline')
  status_parser = subs.add_parser('status', parents=[socket_parser], help='show device status')
  trim_parser = subs.add_parser('trim', parents=[socket_parser], help='request device trim')

  online_subs = online_parser.add_subparsers(required=True)
  offline_subs = offline_parser.add_subparsers(required=True)
  status_subs = status_parser.add_subparsers(required=True)
  trim_subs = trim_parser.add_subparsers(required=True)

  scrub_parser = subs.add_parser("scrub", parents=[socket_parser, cancel_parser], help="request device to be scrubbed")
  scrub_subs = scrub_parser.add_subparsers(required=True)

  list_parser = subs.add_parser("list", parents=[socket_parser, list_args_parser], help="list devices fulfilling conditions")
  list_parser.set_defaults(func=make_list_command(op=None, overdue=False))

  list_subs = list_parser.add_subparsers()
  list_subs.add_parser("overdue", parents=[list_args_parser], help="list all overdue devices").set_defaults(func=make_list_command(op=None, overdue=True, entity_type=ZDev))

  list_scrub_subs = list_subs.add_parser("scrub", help="list devices based on current/last scrub state").add_subparsers(required=True)
  list_scrub_subs.add_parser("overdue", parents=[list_args_parser], help="list devices on which a scrub is overdue").set_defaults(func=make_list_command(Operation.SCRUB, overdue=True))

  list_trim_subs = list_subs.add_parser("trim", help="list devices based on current/last trim state").add_subparsers(required=True)
  list_trim_subs.add_parser("overdue", parents=[list_args_parser], help="list devices on which a trim is overdue").set_defaults(func=make_list_command(Operation.TRIM, overdue=True))

  list_update_subs = list_subs.add_parser("update", help="list devices based on when they were last up-to-date").add_subparsers(required=True)
  list_update_subs.add_parser("overdue", parents=[list_args_parser], help="list devices which need updating of their data").set_defaults(func=make_list_command(Operation.RESILVER, overdue=True))





  enable_subs = subs.add_parser('enable', parents=[socket_parser], help='enable zmirror daemon property').add_subparsers(required=True)
  disable_subs = subs.add_parser('disable', parents=[socket_parser], help='disable zmirror daemon property').add_subparsers(required=True)
  set_subs = subs.add_parser('set', parents=[socket_parser], help='set zmirror daemon property to value').add_subparsers(required=True)
  get_subs = subs.add_parser('get', parents=[socket_parser], help='get zmirror daemon property value').add_subparsers(required=True)

  def add_enable_disable_parsers(name, help=None):
    enable_parser = enable_subs.add_parser(name, help=f"enable {help}")
    disable_parser = disable_subs.add_parser(name, help=f"disable {help}")

    enable_parser.set_defaults(func=make_send_set_property_daemon_command(name, "yes"))
    disable_parser.set_defaults(func=make_send_set_property_daemon_command(name, "no"))


    get_subs.add_parser(name, help=f"get {help}").set_defaults(func=make_send_get_property_daemon_command(name))

  add_enable_disable_parsers("commands", help="command execution")
  add_enable_disable_parsers("log-events", help="logging of all UDEV and ZED events received by zmirror")


  def add_set_property_parser(name, set_help=None, get_help=None):
    set_parser = set_subs.add_parser(name, help=set_help)
    set_parser.add_argument("value", type=str, help="the new value of the property")
    set_parser.set_defaults(func=make_send_set_property_daemon_command(name))
    get_subs.add_parser(name, help=get_help).set_defaults(func=make_send_get_property_daemon_command(name))

  add_set_property_parser("log-level", "temporarily set log level to one of: trace | debug | verbose | info | warning | error | critical")
  add_set_property_parser("timeout", "temporarily set request timeout in seconds")

  def make_onlineable_commands(typ):
    command_name = NAME_FOR_TYPE[typ]

    common_parser = argparse.ArgumentParser(add_help=False)
    for fld in typ.id_fields():
      common_parser.add_argument(f"{fld}", type=str, default=None, help="id field")
    
    online = online_subs.add_parser(command_name, parents=[common_parser, cancel_parser])
    online.set_defaults(func=make_send_request_daemon_command(RequestType.ONLINE, typ))

    if typ in {ZDev}:
      trim = trim_subs.add_parser(command_name, parents=[common_parser, cancel_parser])
      trim.set_defaults(func=make_send_request_daemon_command(RequestType.TRIM, typ))

    offline = offline_subs.add_parser(command_name, parents=[common_parser, cancel_parser])
    offline.set_defaults(func=make_send_request_daemon_command(RequestType.OFFLINE, typ))

    status = status_subs.add_parser(command_name, parents=[common_parser])
    status.set_defaults(func=make_send_entity_daemon_command("status", typ))

    if typ in [ZDev, ZPool]:

      scrub = scrub_subs.add_parser(command_name, parents=[common_parser])
      scrub.set_defaults(func=make_send_request_daemon_command(RequestType.SCRUB, typ))




  for typ in [Disk, Partition, ZPool, ZFSVolume, DMCrypt, ZDev]:
    make_onlineable_commands(typ)



  return parser




def main(args=None):

  parser = make_arg_parser()


  args = parser.parse_args(args=args)

  # if args.cache_path is None:
  #  args.cache_path = args.state_dir + "/cache.yml"

  try:
    args.func(args)
  except Exception as exception:
    error_message = str(exception)
    log.error(error_message)
    traceback.print_exc()



if __name__ == "__main__":
  main()
